test/jdk/java/lang/Thread/crac/JoinOrSleepOnCRPauseTest.java

```diff
@@ -0,0 +1,194 @@
+//
+// This code is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+// A PARTICULAR PURPOSE.  See the GNU General Public License version 2 for more
+// details (a copy is included in the LICENSE file that accompanied this code).
+//
+// You should have received a copy of the GNU General Public License version 2
+// along with this work; if not, write to the Free Software Foundation, Inc.,
+// 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+//
+// Please contact Azul Systems, 385 Moffett Park Drive, Suite 115, Sunnyvale,
+// CA 94089 USA or visit www.azul.com if you need additional information or
+// have any questions.
+
+/*
+ * @test JoinOrSleepOnCRPauseTest.java
+ * @requires (os.family == "linux")
+ * @library /test/lib
+ * @summary check if Thread.join(timeout) and Thread.sleep(timeout)
+ *          will be completed on restore immediately
+ *          if their end time fell on the CRaC pause period
+ *          (i.e. between the checkpoint and restore)
+ * @run main/othervm JoinOrSleepOnCRPauseTest
+ */
+
+
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+
+import java.util.concurrent.CountDownLatch;
+
+
+public class JoinOrSleepOnCRPauseTest {
+
+    private static enum TestType {join, sleep};
+    private final TestType testType;
+
+    private final boolean nsAccuracy;
+
+    private final static long EPS_NS = Long.parseLong(
+        System.getProperty("cractests.time.eps", "100000000")); // default: 0.1s
+
+    private static final long T_MS = 2000;
+    private static final  int T_NS = 100;
+
+    private static final long CRPAUSE_MS = 4000;
+
+    private volatile long tDone = -1;
+
+    private final CountDownLatch checkpointLatch = new CountDownLatch(1);
+
+
+    private JoinOrSleepOnCRPauseTest(TestType testType, boolean nsAccuracy) {
+        this.testType   = testType;
+        this.nsAccuracy = nsAccuracy;
+    }
+
+    private void runTest() throws Exception {
+
+        String op = testType.name();
+
+        Thread mainThread = Thread.currentThread();
+
+        Runnable r = () -> {
+
+            try {
+
+                checkpointLatch.countDown();
+
+                switch (testType) {
+
+                    case join:
+                        if (nsAccuracy) {
+                            mainThread.join(T_MS, T_NS);
+                        } else {
+                            mainThread.join(T_MS);
+                        }
+                        break;
+
+                    case sleep:
+                        if (nsAccuracy) {
+                            Thread.sleep(T_MS, T_NS);
+                        } else {
+                            Thread.sleep(T_MS);
+                        }
+                        break;
+
+                    default:
+                        throw new IllegalArgumentException("unknown test type");
+                }
```

Comment:
FWIW, this can be refactored out as 
```
Runnable sleep_ms = () -> { Thread.sleep(1) }
Runnable sleep_ns = () -> { Thread.sleep(1, 1) }

runTest(sleep_ns)

void runTest(Runnable body) {
  Runnable r = () -> { 
    checkpointLatch.countDown();
    body.run()
  }
}
  
```

Reply:
thanks, will consider this option

Reply:
sorry, after some thougths, i'd prefer to leave this switch. but for now it does not contain nested ifs, making it more transient